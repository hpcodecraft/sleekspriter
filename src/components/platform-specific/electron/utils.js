import { writeFile } from "fs";
import { basename, dirname } from "path";

import config from "../../../config";
import { stateToFile } from "../../../utils";

export const save = (path, contents, errorHandler) => {
  writeFile(path, JSON.stringify(stateToFile(contents)), function(error) {
    if (error) {
      errorHandler(error);
    }
  });
};

export const fileInfoFromPath = path => {
  const fileName = basename(path, "." + config.fileExtension); // the file name without extension, e.g. "coin"
  const folder = dirname(path); // the complete absolute folder (same as path without the file name)

  return {
    fileName,
    folder,
  };
};

export const fileDefaultPath = (file, extension) => {
  const folder = file.folder || config.defaultFolder;
  const name = file.name || config.defaultName;
  const ext = extension || config.fileExtension;

  return `${folder}/${name}.${ext}`; // TODO: use platform-independent folder seperator
};
