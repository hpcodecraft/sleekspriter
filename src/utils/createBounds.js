import { selectionIsActive } from "./";

const createBounds = (size, selection) => {
  if (!selection || !selectionIsActive(selection)) {
    return {
      start: { x: 1, y: 1 },
      end: { x: size.width, y: size.height },
    };
  } else {
    return selection;
  }
};

export default createBounds;
