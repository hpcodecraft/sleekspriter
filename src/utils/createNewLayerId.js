import { maxBy } from "lodash";

const createNewLayerId = layers => {
  return maxBy(layers, "id").id + 1;
};

export default createNewLayerId;
