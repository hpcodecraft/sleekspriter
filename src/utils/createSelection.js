// this function takes two points and returns a selection object where the
// start always is the top left corner and
// end is the bottom right one
// this makes drawing the selection easier and ensures width calculations etc are always consistent

const createSelection = (p1, p2) => {
  return {
    start: {
      x: p1.x < p2.x ? p1.x : p2.x,
      y: p1.y < p2.y ? p1.y : p2.y,
    },
    end: {
      x: p1.x > p2.x ? p1.x : p2.x,
      y: p1.y > p2.y ? p1.y : p2.y,
    },
  };
};

export default createSelection;
