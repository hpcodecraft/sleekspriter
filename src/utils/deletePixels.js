const deletePixels = (allPixels, frame, layer, pixelsToDelete) => {

  Object.keys(pixelsToDelete).map(x => {
    Object.keys(pixelsToDelete[x]).map(y => {
      // console.log("delete! ", frame, layer, x, y);
      delete allPixels[frame][layer][x][y];
    });

    if(Object.keys(allPixels[frame][layer][x]).length === 0) {
      // console.log("delete! ", frame, layer, x);
      delete allPixels[frame][layer][x];
    }
  });

  if(Object.keys(allPixels[frame][layer]).length === 0) {
    // console.log("delete! ", frame, layer);
    delete allPixels[frame][layer];
  }
  if(Object.keys(allPixels[frame]).length === 0) {
    // console.log("delete! ", frame);
    delete allPixels[frame];
  }

  return allPixels;
};

export default deletePixels;
