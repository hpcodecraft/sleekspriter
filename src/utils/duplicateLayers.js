const duplicateLayers = (layers, targetFrame, nextLayerId) => {
  const newLayers = [];
  for (let i = 0; i < layers.length; i++) {
    const newId = nextLayerId + i;
    const newLayer = Object.assign({}, layers[i]);
    newLayer.id = newId;
    newLayer.frame = targetFrame;
    newLayer.name = `${newLayer.name} copy`;
    newLayers.push(newLayer);
  }

  return newLayers;
};

export default duplicateLayers;
