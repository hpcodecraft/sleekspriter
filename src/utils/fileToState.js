// converts JSON from file format to state-ready JSON
// TODO: add animations

const fileToState = file => {
  let { frames, layers, pixels, size } = file;

  layers = layers.map(layerToState);

  const frameDict = {};
  layers.forEach(layer => {
    frameDict[layer.id] = layer.frame;
  });

  const mappedPixels = {};
  pixels.forEach(p => {
    const pixel = {
      frame: frameDict[p[0]],
      layer: p[0],
      x: p[1],
      y: p[2],
      r: p[3],
      g: p[4],
      b: p[5],
      a: p[6],
    };

    if (!mappedPixels[pixel.frame]) mappedPixels[pixel.frame] = {};
    if (!mappedPixels[pixel.frame][pixel.layer])
      mappedPixels[pixel.frame][pixel.layer] = {};
    if (!mappedPixels[pixel.frame][pixel.layer][pixel.x])
      mappedPixels[pixel.frame][pixel.layer][pixel.x] = {};
    mappedPixels[pixel.frame][pixel.layer][pixel.x][pixel.y] = pixel;
  });

  return {
    frames: {
      x: frames[0],
      y: frames[1],
    },
    layers,
    pixels: mappedPixels,
    size: {
      width: size[0],
      height: size[1],
    },
  };
};

const layerToState = layer => ({
  id: layer[0],
  frame: layer[1],
  name: layer[2],
  z: layer[3],
  opacity: layer[4],
  visible: !!layer[5],
});

export default fileToState;
