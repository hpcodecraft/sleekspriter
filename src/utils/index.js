export capitalizeFirstLetter from "./capitalizeFirstLetter";
export createBounds from "./createBounds";
export createNewLayerId from "./createNewLayerId";
export deletePixels from "./deletePixels";
export duplicateLayers from "./duplicateLayers";
export flattenPixels from "./flattenPixels";
export getPivot from "./getPivot";
export getPixelsInFrame from "./getPixelsInFrame";
export getPixelsInScope from "./getPixelsInScope";
export fileToState from "./fileToState";
export insideBounds from "./insideBounds";
export mergeLayerPixels from "./mergeLayerPixels";
export selectionIsActive from "./selectionIsActive";
export stateToFile from "./stateToFile";
export t from "./translate";
