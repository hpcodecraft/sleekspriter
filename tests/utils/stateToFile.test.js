jest.unmock("utils");

import { stateToFile } from "utils";

describe("stateToFile", () => {
  it("maps JSON from redux state to file format", () => {
    const state = {
      frames: {
        x: 5,
        y: 3
      },
      layers: [
        { frame: 1, id: 2, name: "Layer 2", z: 1, opacity: 100, visible: true },
        { frame: 1, id: 1, name: "Layer 1", z: 0, opacity: 100, visible: true },
        { frame: 1, id: 3, name: "Layer 3", z: 2, opacity: 100, visible: true },
        { frame: 2, id: 4, name: "Layer 4", z: 0, opacity: 100, visible: true },
        { frame: 3, id: 5, name: "Layer 5", z: 0, opacity: 100, visible: true },
        { frame: 4, id: 6, name: "Layer 6", z: 0, opacity: 100, visible: true },
        { frame: 5, id: 7, name: "Layer 7", z: 0, opacity: 100, visible: true },
        { frame: 6, id: 8, name: "Layer 8", z: 0, opacity: 100, visible: true }
      ],
      pixels: {
        1: {
          1: {
            1: {
              1: { frame: 1, layer: 1, x: 1, y: 1, r: 0, g: 0, b: 0, a: 1 }
            }
          }
        }
      },
      size: {
        height: 32,
        width: 32
      }
    };

    expect(stateToFile(state)).toMatchSnapshot();
  });
});
