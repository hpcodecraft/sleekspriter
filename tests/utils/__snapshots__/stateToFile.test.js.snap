// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`stateToFile maps JSON from redux state to file format 1`] = `
Object {
  "frames": Array [
    5,
    3,
  ],
  "layers": Array [
    Array [
      2,
      1,
      "Layer 2",
      1,
      100,
      1,
    ],
    Array [
      1,
      1,
      "Layer 1",
      0,
      100,
      1,
    ],
    Array [
      3,
      1,
      "Layer 3",
      2,
      100,
      1,
    ],
    Array [
      4,
      2,
      "Layer 4",
      0,
      100,
      1,
    ],
    Array [
      5,
      3,
      "Layer 5",
      0,
      100,
      1,
    ],
    Array [
      6,
      4,
      "Layer 6",
      0,
      100,
      1,
    ],
    Array [
      7,
      5,
      "Layer 7",
      0,
      100,
      1,
    ],
    Array [
      8,
      6,
      "Layer 8",
      0,
      100,
      1,
    ],
  ],
  "meta": Object {
    "author": "John Doe",
    "created": "1970-01-01T00:00:00.000Z",
    "homepage": "https://example.com",
    "updated": "1970-01-01T00:00:00.000Z",
    "version": 1,
  },
  "pixels": Array [
    Array [
      1,
      1,
      1,
      0,
      0,
      0,
      1,
    ],
  ],
  "size": Array [
    32,
    32,
  ],
}
`;
